### DISCUSSION TITLE
Built-ins with Web Workers

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1322] [Q&A] [answered]
What is the recommended way to access a code file for loading a set of built-ins using a custom WorkspaceManager when you are using WebWorkers?  My guess is that the NodeFileSystemProvider won't work for me if I am bundling everything for execution in the browser instead of running my language server as a NodeJS app.

Here is how I have done it in that past with a standard NodeJS implementation:

```typescript
protected override async loadAdditionalDocuments(_folders: WorkspaceFolder[], _collector: (document: LangiumDocument) => void): Promise<void> {
      
        //read in the global behaviors file
        const fileSystemProvider = new NodeFileSystemProvider();
        const uri: URI = URI.file(__dirname + '/resources/global.behaviors') ; // swiped from https://github.com/microsoft/vscode-uri
        const text = await fileSystemProvider.readFile(uri);

        //parse the doc contents
        const doc : LangiumDocument = this.factory.fromString(text,uri);
        _collector(doc);

        // add doc to the cache, as suggested in https://github.com/langium/langium/issues/464
        this.langiumDocuments.addDocument(doc);
        return Promise.resolve();

    }
```

**Comment by msujew** [comment, marked as answer]
@ballcoach12 Often times we usually just put the builtin library into the TypeScript code:

```ts
export const stdlib = `
// stdlib here
`;
```

If you don't want to embed it into the bundle itself, you will likely need to download it from some service endpoint. If you have both a node and web version of your language, it makes sense to have something like a `BuiltInLoadStrategy` as part of your injection context that you pass to your `create<language>Module` function. One would just load from disk using `fs` while the other one uses `fetch` to download the builtin library instead.

**Reply by ballcoach12** [reply]
Thanks, @msujew. I was heading in that direction but wanted to make sure there wasn't a better way before I wrote the code. 

I got my import scope mechanism working, which downloads from a remote endpoint. But that approach might be overkill for built-ins.

**Reply by cdietrich** [reply]
We open currently experiment with two strategies
- warm up the webworker on an initial message call and send the otherFiles there and create a languim built in file aware module that knows about the specific file content sent to the webworker
- simply open more than 1 file from the client so that the server sees multiple files from the main document
